pkg/volume/csimigration/plugin_manager.go

```diff
@@ -0,0 +1,154 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package csimigration
+
+import (
+	"errors"
+	"fmt"
+
+	"k8s.io/api/core/v1"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	"k8s.io/component-base/featuregate"
+	csilibplugins "k8s.io/csi-translation-lib/plugins"
+	"k8s.io/kubernetes/pkg/features"
+	"k8s.io/kubernetes/pkg/volume"
+)
+
+// PluginNameMapper contains utility methods to retrieve names of plugins
+// that support a spec, map intree <=> migrated CSI plugin names, etc
+type PluginNameMapper interface {
+	GetInTreePluginNameFromSpec(pv *v1.PersistentVolume, vol *v1.Volume) (string, error)
+	GetCSINameFromInTreeName(pluginName string) (string, error)
+}
+
+// PluginManager keeps track of migrated state of in-tree plugins
+type PluginManager struct {
+	PluginNameMapper
+}
+
+// NewPluginManager returns a new PluginManager instance
+func NewPluginManager(m PluginNameMapper) PluginManager {
+	return PluginManager{
+		PluginNameMapper: m,
+	}
+}
+
+// IsMigrationCompleteForPlugin indicates whether CSI migration has been completed
+// for a particular storage plugin
+func (pm PluginManager) IsMigrationCompleteForPlugin(pluginName string) bool {
```

Comment:
nothing in this method uses `pm`... this could be a function instead of a method, then be used by kube-controller-manager and kubelet so they don't have to maintain their own plugin -> feature name mapping

Reply:
@liggitt I completely agree that in the context of the above, `IsMigrationCompleteForPlugin`/`IsMigrationEnabledForPlugin` could be a function instead of method. However, the reason I did not make it a function is: enabling the unit testing of modules using these methods (such as the pv_controller) and align with the overall pattern introduced in https://github.com/kubernetes/kubernetes/pull/82683. 

So having these as methods allows an easy way for unit tests to create an interface implementing `IsMigrationCompleteForPlugin`/`IsMigrationEnabledForPlugin` and use fake plugins (for example for [pv_controller](https://github.com/kubernetes/kubernetes/pull/83098/files#diff-6c9081ca34ca47769f31d0df75a50934R532)).

Reply:
having the feature mapping in multiple places is going to be hard to maintain if more things are added to migration (especially if implementations fall back to reporting false if an unknown plugin is asked about). I still think these should collapse to a single authoritative spot in the code, but that can be a follow up. David opened https://github.com/kubernetes/kubernetes/issues/85313 to track that.

Reply:
Totally agree about single spot in the code for the mappings rather than multiple places.